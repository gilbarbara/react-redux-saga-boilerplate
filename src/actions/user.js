// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { userLogin: login, userLogout: logOut } = createActions({
  [ActionTypes.USER_LOGIN]: () => ({}),
  [ActionTypes.USER_LOGOUT]: () => ({}),
});
