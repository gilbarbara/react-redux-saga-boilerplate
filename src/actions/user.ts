/**
 * @module Actions/User
 * @desc User Actions
 */
import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

export const login = createAction(ActionTypes.USER_LOGIN_REQUEST, () => ({}));
export const logOut = createAction(ActionTypes.USER_LOGOUT_REQUEST, () => ({}));
