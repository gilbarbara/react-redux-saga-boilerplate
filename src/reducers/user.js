import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';

import { STATUS, ActionTypes } from 'constants/index';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGIN_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: true },
          status: { $set: STATUS.READY },
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          status: { $set: STATUS.IDLE },
        }),
    },
    userState,
  ),
};
