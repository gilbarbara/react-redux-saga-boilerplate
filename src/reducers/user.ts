import { createReducer } from 'modules/helpers';

import { ActionTypes, STATUS } from 'literals';

import { UserState } from 'types';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: createReducer<UserState>(
    {
      [ActionTypes.USER_LOGIN_REQUEST]: draft => {
        draft.status = STATUS.RUNNING;
      },
      [ActionTypes.USER_LOGIN_SUCCESS]: draft => {
        draft.isAuthenticated = true;
        draft.status = STATUS.READY;
      },
      [ActionTypes.USER_LOGOUT_REQUEST]: draft => {
        draft.status = STATUS.RUNNING;
      },
      [ActionTypes.USER_LOGOUT_SUCCESS]: draft => {
        draft.isAuthenticated = false;
        draft.status = STATUS.IDLE;
      },
    },
    userState,
  ),
};
