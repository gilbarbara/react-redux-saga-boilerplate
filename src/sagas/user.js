/**
 * @module Sagas/User
 * @desc User
 */

import { delay } from 'redux-saga';
import { all, call, put, takeLatest } from 'redux-saga/effects';

import { ActionTypes } from 'constants/index';

/**
 * Login
 */
export function* login() {
  try {
    yield call(delay, 400);

    yield put({
      type: ActionTypes.USER_LOGIN_SUCCESS,
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN_FAILURE,
      payload: err,
    });
  }
}

/**
 * Logout
 */
export function* logout() {
  try {
    yield call(delay, 200);

    yield put({
      type: ActionTypes.USER_LOGOUT_SUCCESS,
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGOUT_FAILURE,
      payload: err,
    });
  }
}

/**
 * User Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN, login),
    takeLatest(ActionTypes.USER_LOGOUT, logout),
  ]);
}
