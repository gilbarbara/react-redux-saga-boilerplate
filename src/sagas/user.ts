/**
 * @module Sagas/User
 * @desc User
 */

import { all, delay, put, takeLatest } from 'redux-saga/effects';

import { ActionTypes } from 'literals';

/**
 * Login
 */
export function* login() {
  yield delay(400);

  yield put({
    type: ActionTypes.USER_LOGIN_SUCCESS,
  });
}

/**
 * Logout
 */
export function* logout() {
  yield delay(200);

  yield put({
    type: ActionTypes.USER_LOGOUT_SUCCESS,
  });
}

/**
 * User Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN_REQUEST, login),
    takeLatest(ActionTypes.USER_LOGOUT_REQUEST, logout),
  ]);
}
