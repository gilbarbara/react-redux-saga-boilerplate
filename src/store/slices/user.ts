import { createSlice } from '@reduxjs/toolkit';

import { STATUS } from '~/literals';

import { UserState } from '~/types';

export const userState: UserState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export const userSlice = createSlice({
  name: 'user',
  initialState: userState,
  reducers: {
    login: draft => {
      draft.status = STATUS.RUNNING;
    },
    loginSuccess: draft => {
      draft.isAuthenticated = true;
      draft.status = STATUS.READY;
    },
    logOut: draft => {
      draft.status = STATUS.RUNNING;
    },
    logOutSuccess: draft => {
      draft.isAuthenticated = false;
      draft.status = STATUS.IDLE;
    },
  },
});

export const { login, loginSuccess, logOut, logOutSuccess } = userSlice.actions;
export default userSlice.reducer;
