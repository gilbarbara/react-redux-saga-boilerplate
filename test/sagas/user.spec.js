import user, { watchLogin, watchLogout } from 'sagas/user';
import { ActionTypes } from 'constants/index';

describe('user', () => {
  it('should have the expected watchers', () => {
    const saga = user();
    const watchers = saga.next().value.ALL;

    expect(watchers).toHaveLength(2);
    expect(watchers).toMatchSnapshot();
    expect(saga.next().done).toBe(true);
  });

  it('should match the login saga', () => {
    const watcher = watchLogin();
    expect(watcher.next().value).toMatchSnapshot();

    const loginCall = watcher.next().value;
    expect(watcher.next().done).toBe(true);
    expect(typeof loginCall.CALL.fn).toBe('function');

    const login = loginCall.CALL.fn();
    expect(login.next().value.CALL.args).toEqual([400]);
    expect(login.next().value.PUT.action.type).toBe(ActionTypes.USER_LOGIN_SUCCESS);
    expect(login.next().value.PUT.action.type).toEqual('@@router/CALL_HISTORY_METHOD');
    expect(login.next().done).toBe(true);
  });

  it('should match the logout saga', () => {
    const watcher = watchLogout();
    expect(watcher.next().value).toMatchSnapshot();

    const logoutCall = watcher.next().value;
    expect(watcher.next().done).toBe(true);
    expect(typeof logoutCall.CALL.fn).toBe('function');

    const logout = logoutCall.CALL.fn();
    expect(logout.next().value.CALL.args).toEqual([200]);
    expect(logout.next().value.PUT.action.type).toBe(ActionTypes.USER_LOGOUT_SUCCESS);
    expect(logout.next().value.PUT.action.type).toEqual('@@router/CALL_HISTORY_METHOD');
    expect(logout.next().done).toBe(true);
  });
});
