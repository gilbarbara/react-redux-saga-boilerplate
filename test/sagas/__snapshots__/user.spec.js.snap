// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user should have the expected watchers 1`] = `
Array [
  Object {
    "@@redux-saga/IO": true,
    "FORK": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
  Object {
    "@@redux-saga/IO": true,
    "FORK": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
]
`;

exports[`user should match the login saga 1`] = `
Object {
  "@@redux-saga/IO": true,
  "TAKE": Object {
    "pattern": "USER_LOGIN_REQUEST",
  },
}
`;

exports[`user should match the logout saga 1`] = `
Object {
  "@@redux-saga/IO": true,
  "TAKE": Object {
    "pattern": "USER_LOGOUT_REQUEST",
  },
}
`;
